`pants check` fails on Terraform modules due to lack of `terraform init`

**Describe the bug**
After setting up some basic Terraform in my repo, running `pants check ::` fails with the following error message:
```
16:48:40.39 [ERROR] Completed: pants.backend.terraform.goals.check.terraform_check - terraform-validate failed (exit code 1).
Partition #1 - `terraform validate` on `myrepo/terraform/environments/dev`:
╷
│ Error: Module not installed
│ 
│   on myrepo/terraform/environments/dev/main.tf line 23:
│   23: module "my_module" {
│ 
│ This module is not yet installed. Run "terraform init" to install all
│ modules required by this configuration.
╵

Partition #2 - `terraform validate` on `myrepo/terraform/modules/infrastructure`:
╷
│ Error: Missing required provider
│ 
│ This configuration requires provider
│ registry.terraform.io/hashicorp/azurerm, but that provider isn't available.
│ You may be able to install it automatically by running:
│   terraform init
╵



✕ terraform-validate failed.
```

**Pants version**
2.15

**OS**
Intel Mac

**Additional info**
Ideally, `pants check` would run `init`, `validate`, and `plan`. I'm going to open a feature request for this exact thing. It would also be super useful if you could run `terraform apply` through Pants, either using the `publish` or `experimental-deploy` goals (another issue, the difference between these two should be elaborated on as they're both from the core backend).

--- a/src/python/pants/backend/experimental/terraform/register.py
+++ b/src/python/pants/backend/experimental/terraform/register.py
@@ -2,7 +2,6 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
 from pants.backend.python.goals import lockfile as python_lockfile
-from pants.backend.python.util_rules.pex import rules as pex_rules
 from pants.backend.terraform import dependency_inference, tool
 from pants.backend.terraform.goals import check, tailor
 from pants.backend.terraform.lint.tffmt.tffmt import rules as tffmt_rules
@@ -23,7 +22,6 @@ def rules():
         *tailor.rules(),
         *target_types_rules(),
         *tool.rules(),
-        *pex_rules(),
         *tffmt_rules(),
         *python_lockfile.rules(),
     ]

--- a/src/python/pants/backend/terraform/dependency_inference.py
+++ b/src/python/pants/backend/terraform/dependency_inference.py
@@ -2,9 +2,8 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 from __future__ import annotations
 
-import shlex
 from dataclasses import dataclass
-from pathlib import Path, PurePath
+from pathlib import PurePath
 from typing import Tuple
 
 from pants.backend.python.subsystems.python_tool_base import (
@@ -167,11 +166,12 @@ async def get_terraform_providers(
         Get(
             FallibleProcessResult,
             TerraformProcess(
-                args=(f"-chdir={shlex.quote(directory)}", "init"),
+                args=("init",),
                 input_digest=req.source_files.snapshot.digest,
-                output_files=((Path(directory) / ".terraform.lock.hcl").as_posix(),),
-                output_directories=((Path(directory) / ".terraform").as_posix(),),
+                output_files=(".terraform.lock.hcl",),
+                output_directories=(".terraform",),
                 description="Run `terraform init` to fetch dependencies",
+                chdir=directory,
             ),
         )
         for directory in req.directories

--- a/src/python/pants/backend/terraform/dependency_inference.py
+++ b/src/python/pants/backend/terraform/dependency_inference.py
@@ -13,11 +13,13 @@ from pants.backend.python.target_types import EntryPoint
 from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
 from pants.backend.python.util_rules.pex import rules as pex_rules
 from pants.backend.terraform.target_types import TerraformModuleSourcesField
+from pants.backend.terraform.tool import TerraformProcess
 from pants.base.glob_match_error_behavior import GlobMatchErrorBehavior
 from pants.base.specs import DirGlobSpec, RawSpecs
+from pants.core.util_rules.source_files import SourceFiles
 from pants.engine.fs import CreateDigest, Digest, FileContent
 from pants.engine.internals.selectors import Get
-from pants.engine.process import Process, ProcessResult
+from pants.engine.process import FallibleProcessResult, Process, ProcessResult
 from pants.engine.rules import collect_rules, rule
 from pants.engine.target import (
     FieldSet,
@@ -142,6 +144,33 @@ async def infer_terraform_module_dependencies(
         tgt.address for tgt in candidate_targets if tgt.has_field(TerraformModuleSourcesField)
     ]
     return InferredDependencies(terraform_module_addresses)
+
+
+@dataclass(frozen=True)
+class GetTerraformDependenciesRequest:
+    source_files: SourceFiles
+
+
+@dataclass(frozen=True)
+class TerraformDependencies:
+    fetched_deps: Digest
+
+
+@rule
+async def get_terraform_providers(
+    req: GetTerraformDependenciesRequest,
+) -> TerraformDependencies:
+    fetched_deps = await Get(
+        FallibleProcessResult,
+        TerraformProcess(
+            args=("init",),
+            input_digest=req.source_files.snapshot.digest,
+            output_files=(".terraform.lock.hcl",),
+            output_directories=(".terraform",),
+            description="Run `terraform init` to fetch dependencies",
+        ),
+    )
+    return TerraformDependencies(fetched_deps.output_digest)
 
 
 def rules():

--- a/src/python/pants/backend/terraform/dependency_inference.py
+++ b/src/python/pants/backend/terraform/dependency_inference.py
@@ -11,6 +11,7 @@ from pants.backend.python.subsystems.python_tool_base import (
 )
 from pants.backend.python.target_types import EntryPoint
 from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.backend.python.util_rules.pex import rules as pex_rules
 from pants.backend.terraform.target_types import TerraformModuleSourcesField
 from pants.base.glob_match_error_behavior import GlobMatchErrorBehavior
 from pants.base.specs import DirGlobSpec, RawSpecs
@@ -146,5 +147,6 @@ async def infer_terraform_module_dependencies(
 def rules():
     return [
         *collect_rules(),
+        *pex_rules(),
         UnionRule(InferDependenciesRequest, InferTerraformModuleDependenciesRequest),
     ]

--- a/src/python/pants/backend/terraform/dependency_inference.py
+++ b/src/python/pants/backend/terraform/dependency_inference.py
@@ -2,8 +2,10 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 from __future__ import annotations
 
+import shlex
 from dataclasses import dataclass
-from pathlib import PurePath
+from pathlib import Path, PurePath
+from typing import Tuple
 
 from pants.backend.python.subsystems.python_tool_base import (
     LockfileRules,
@@ -18,7 +20,7 @@ from pants.base.glob_match_error_behavior import GlobMatchErrorBehavior
 from pants.base.specs import DirGlobSpec, RawSpecs
 from pants.core.util_rules.source_files import SourceFiles
 from pants.engine.fs import CreateDigest, Digest, FileContent
-from pants.engine.internals.selectors import Get
+from pants.engine.internals.selectors import Get, MultiGet
 from pants.engine.process import FallibleProcessResult, Process, ProcessResult
 from pants.engine.rules import collect_rules, rule
 from pants.engine.target import (
@@ -149,28 +151,35 @@ async def infer_terraform_module_dependencies(
 @dataclass(frozen=True)
 class GetTerraformDependenciesRequest:
     source_files: SourceFiles
+    directories: Tuple[str, ...]
 
 
 @dataclass(frozen=True)
 class TerraformDependencies:
-    fetched_deps: Digest
+    fetched_deps: Tuple[Tuple[str, Digest], ...]
 
 
 @rule
 async def get_terraform_providers(
     req: GetTerraformDependenciesRequest,
 ) -> TerraformDependencies:
-    fetched_deps = await Get(
-        FallibleProcessResult,
-        TerraformProcess(
-            args=("init",),
-            input_digest=req.source_files.snapshot.digest,
-            output_files=(".terraform.lock.hcl",),
-            output_directories=(".terraform",),
-            description="Run `terraform init` to fetch dependencies",
-        ),
+    fetched_deps = await MultiGet(
+        Get(
+            FallibleProcessResult,
+            TerraformProcess(
+                args=(f"-chdir={shlex.quote(directory)}", "init"),
+                input_digest=req.source_files.snapshot.digest,
+                output_files=((Path(directory) / ".terraform.lock.hcl").as_posix(),),
+                output_directories=((Path(directory) / ".terraform").as_posix(),),
+                description="Run `terraform init` to fetch dependencies",
+            ),
+        )
+        for directory in req.directories
     )
-    return TerraformDependencies(fetched_deps.output_digest)
+
+    return TerraformDependencies(
+        tuple(zip(req.directories, (x.output_digest for x in fetched_deps)))
+    )
 
 
 def rules():

--- a/src/python/pants/backend/terraform/dependency_inference_test.py
+++ b/src/python/pants/backend/terraform/dependency_inference_test.py
@@ -4,7 +4,6 @@ import textwrap
 
 import pytest
 
-from pants.backend.python.util_rules.pex import rules as pex_rules
 from pants.backend.terraform import dependency_inference
 from pants.backend.terraform.dependency_inference import (
     InferTerraformModuleDependenciesRequest,
@@ -36,7 +35,6 @@ def rule_runner() -> RuleRunner:
         rules=[
             *external_tool.rules(),
             *source_files.rules(),
-            *pex_rules(),
             *dependency_inference.rules(),
             QueryRule(InferredDependencies, [InferTerraformModuleDependenciesRequest]),
             QueryRule(HydratedSources, [HydrateSourcesRequest]),

--- a/src/python/pants/backend/terraform/goals/check.py
+++ b/src/python/pants/backend/terraform/goals/check.py
@@ -7,6 +7,7 @@ from pants.backend.terraform.target_types import TerraformFieldSet
 from pants.backend.terraform.tool import TerraformProcess
 from pants.core.goals.check import CheckRequest, CheckResult, CheckResults
 from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.internals.native_engine import Digest, MergeDigests
 from pants.engine.internals.selectors import Get, MultiGet
 from pants.engine.process import FallibleProcessResult
 from pants.engine.rules import collect_rules, rule
@@ -41,12 +42,27 @@ async def terraform_check(
     )
     files_by_directory = partition_files_by_directory(source_files.files)
 
+    fetched_deps = await Get(
+        FallibleProcessResult,
+        TerraformProcess(
+            args=("init",),
+            input_digest=source_files.snapshot.digest,
+            output_files=(".terraform.lock.hcl",),
+            output_directories=(".terraform",),
+            description="Run `terraform init` to fetch dependencies",
+        ),
+    )
+
+    sources_and_deps = await Get(
+        Digest, MergeDigests([source_files.snapshot.digest, fetched_deps.output_digest])
+    )
+
     results = await MultiGet(
         Get(
             FallibleProcessResult,
             TerraformProcess(
                 args=("validate", directory),
-                input_digest=source_files.snapshot.digest,
+                input_digest=sources_and_deps,
                 output_files=tuple(files),
                 description=f"Run `terraform fmt` on {pluralize(len(files), 'file')}.",
             ),

--- a/src/python/pants/backend/terraform/goals/check.py
+++ b/src/python/pants/backend/terraform/goals/check.py
@@ -1,5 +1,7 @@
 # Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
+import shlex
+
 from pants.backend.terraform.dependency_inference import (
     GetTerraformDependenciesRequest,
     TerraformDependencies,
@@ -44,17 +46,22 @@ async def terraform_check(
     )
     files_by_directory = partition_files_by_directory(source_files.files)
 
-    fetched_deps = await Get(TerraformDependencies, GetTerraformDependenciesRequest(source_files))
+    fetched_deps = await Get(
+        TerraformDependencies,
+        GetTerraformDependenciesRequest(source_files, tuple(files_by_directory.keys())),
+    )
+    # just merge them all for now. This will probably be a problem with multiple TF sources requesting different versions of the same providers
+    merged_fetched_deps = await Get(Digest, MergeDigests([x[1] for x in fetched_deps.fetched_deps]))
 
     sources_and_deps = await Get(
-        Digest, MergeDigests([source_files.snapshot.digest, fetched_deps.fetched_deps])
+        Digest, MergeDigests([source_files.snapshot.digest, merged_fetched_deps])
     )
 
     results = await MultiGet(
         Get(
             FallibleProcessResult,
             TerraformProcess(
-                args=("validate", directory),
+                args=(f"-chdir={shlex.quote(directory)}", "validate"),
                 input_digest=sources_and_deps,
                 output_files=tuple(files),
                 description=f"Run `terraform fmt` on {pluralize(len(files), 'file')}.",

--- a/src/python/pants/backend/terraform/goals/check.py
+++ b/src/python/pants/backend/terraform/goals/check.py
@@ -1,6 +1,5 @@
 # Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
-import shlex
 
 from pants.backend.terraform.dependency_inference import (
     GetTerraformDependenciesRequest,
@@ -61,10 +60,11 @@ async def terraform_check(
         Get(
             FallibleProcessResult,
             TerraformProcess(
-                args=(f"-chdir={shlex.quote(directory)}", "validate"),
+                args=("validate",),
                 input_digest=sources_and_deps,
                 output_files=tuple(files),
                 description=f"Run `terraform fmt` on {pluralize(len(files), 'file')}.",
+                chdir=directory,
             ),
         )
         for directory, files in files_by_directory.items()

--- a/src/python/pants/backend/terraform/goals/check.py
+++ b/src/python/pants/backend/terraform/goals/check.py
@@ -1,7 +1,9 @@
 # Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
-
-
+from pants.backend.terraform.dependency_inference import (
+    GetTerraformDependenciesRequest,
+    TerraformDependencies,
+)
 from pants.backend.terraform.partition import partition_files_by_directory
 from pants.backend.terraform.target_types import TerraformFieldSet
 from pants.backend.terraform.tool import TerraformProcess
@@ -42,19 +44,10 @@ async def terraform_check(
     )
     files_by_directory = partition_files_by_directory(source_files.files)
 
-    fetched_deps = await Get(
-        FallibleProcessResult,
-        TerraformProcess(
-            args=("init",),
-            input_digest=source_files.snapshot.digest,
-            output_files=(".terraform.lock.hcl",),
-            output_directories=(".terraform",),
-            description="Run `terraform init` to fetch dependencies",
-        ),
-    )
+    fetched_deps = await Get(TerraformDependencies, GetTerraformDependenciesRequest(source_files))
 
     sources_and_deps = await Get(
-        Digest, MergeDigests([source_files.snapshot.digest, fetched_deps.output_digest])
+        Digest, MergeDigests([source_files.snapshot.digest, fetched_deps.fetched_deps])
     )
 
     results = await MultiGet(

--- a/src/python/pants/backend/terraform/goals/check_test.py
+++ b/src/python/pants/backend/terraform/goals/check_test.py
@@ -155,3 +155,22 @@ def test_with_dependency(rule_runner: RuleRunner) -> None:
     targets = [make_target(rule_runner, [SOURCE_WITH_PROVIDER])]
     check_results = run_terraform_validate(rule_runner, targets)
     assert check_results[0].exit_code == 0
+
+
+def test_in_folder(rule_runner: RuleRunner) -> None:
+    """Test that we can `check` terraform files not in the root folder."""
+    target_name = "in_folder"
+    files = {
+        "folder/BUILD": f"terraform_module(name='{target_name}')\n",
+        "folder/provided.tf": textwrap.dedent(
+            """
+            resource "null_resource" "dep" {}
+            resource "random_pet" "random" {}
+            """
+        ),
+    }
+    rule_runner.write_files(files)
+    target = rule_runner.get_target(Address("folder", target_name=target_name))
+
+    check_results = run_terraform_validate(rule_runner, [target])
+    assert check_results[0].exit_code == 0

--- a/src/python/pants/backend/terraform/goals/check_test.py
+++ b/src/python/pants/backend/terraform/goals/check_test.py
@@ -3,7 +3,7 @@
 from __future__ import annotations
 
 import textwrap
-from typing import Sequence
+from typing import Dict, Sequence
 
 import pytest
 
@@ -174,3 +174,42 @@ def test_in_folder(rule_runner: RuleRunner) -> None:
 
     check_results = run_terraform_validate(rule_runner, [target])
     assert check_results[0].exit_code == 0
+
+
+def test_conflicting_provider_versions(rule_runner: RuleRunner) -> None:
+    """Test that 2 separate terraform_modules can request conflicting providers."""
+    target_name = "in_folder"
+    versions = ["3.2.1", "3.0.0"]
+
+    def make_terraform_module(version: str) -> Dict[str, str]:
+        return {
+            f"folder{version}/BUILD": f"terraform_module(name='{target_name}')\n",
+            f"folder{version}/provided.tf": textwrap.dedent(
+                """
+            terraform {
+              required_providers {
+                null = {
+                  source = "hashicorp/null"
+                  version = "%s"
+                }
+              }
+            }
+            resource "null_resource" "res" {}
+            """
+                % version
+            ),
+        }
+
+    files = {}
+    for version in versions:
+        files.update(make_terraform_module(version))
+
+    rule_runner.write_files(files)
+    targets = [
+        rule_runner.get_target(Address(folder, target_name=target_name))
+        for folder in (f"folder{version}" for version in versions)
+    ]
+
+    check_results = run_terraform_validate(rule_runner, targets)
+    assert len(check_results) == len(versions)
+    assert all(check_result.exit_code == 0 for check_result in check_results)

--- a/src/python/pants/backend/terraform/goals/check_test.py
+++ b/src/python/pants/backend/terraform/goals/check_test.py
@@ -7,7 +7,7 @@ from typing import Sequence
 
 import pytest
 
-from pants.backend.terraform import tool
+from pants.backend.terraform import dependency_inference, tool
 from pants.backend.terraform.goals import check
 from pants.backend.terraform.goals.check import TerraformCheckRequest
 from pants.backend.terraform.target_types import TerraformFieldSet, TerraformModuleTarget
@@ -29,6 +29,7 @@ def rule_runner() -> RuleRunner:
             *check.rules(),
             *tool.rules(),
             *source_files.rules(),
+            *dependency_inference.rules(),
             QueryRule(CheckResults, (TerraformCheckRequest,)),
             QueryRule(SourceFiles, (SourceFilesRequest,)),
         ],

--- a/src/python/pants/backend/terraform/goals/check_test.py
+++ b/src/python/pants/backend/terraform/goals/check_test.py
@@ -60,6 +60,16 @@ BAD_SOURCE = FileContent(
     ).encode("utf-8"),
 )
 
+# This resource uses the null_resource provider. Terraform will need to run `init` to init the provider
+SOURCE_WITH_PROVIDER = FileContent(
+    "provided.tf",
+    textwrap.dedent(
+        """
+        resource "null_resource" "dep" {}
+        """
+    ).encode("utf-8"),
+)
+
 
 def make_target(
     rule_runner: RuleRunner, source_files: list[FileContent], *, target_name="target"
@@ -133,3 +143,14 @@ def test_skip(rule_runner: RuleRunner) -> None:
     target = make_target(rule_runner, [BAD_SOURCE])
     lint_results = run_terraform_validate(rule_runner, [target], args=["--terraform-validate-skip"])
     assert not lint_results
+
+
+def test_with_dependency(rule_runner: RuleRunner) -> None:
+    """Sources with a provider need to have `terraform init` run before to initialise the provider.
+
+    Without `init`, `terraform validate` fails. It is therefore sufficient to just test that the
+    process ran successfully
+    """
+    targets = [make_target(rule_runner, [SOURCE_WITH_PROVIDER])]
+    check_results = run_terraform_validate(rule_runner, targets)
+    assert check_results[0].exit_code == 0

--- a/src/python/pants/backend/terraform/tool.py
+++ b/src/python/pants/backend/terraform/tool.py
@@ -2,7 +2,10 @@
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 from __future__ import annotations
 
+import shlex
 from dataclasses import dataclass
+from pathlib import Path
+from typing import Tuple
 
 from pants.core.util_rules.external_tool import (
     DownloadedExternalTool,
@@ -186,6 +189,7 @@ class TerraformProcess:
     input_digest: Digest = EMPTY_DIGEST
     output_files: tuple[str, ...] = ()
     output_directories: tuple[str, ...] = ()
+    chdir: str = "."  # directory for terraform's `-chdir` argument
 
 
 @rule
@@ -200,12 +204,15 @@ async def setup_terraform_process(
 
     immutable_input_digests = {"__terraform": downloaded_terraform.digest}
 
+    def prepend_paths(paths: Tuple[str, ...]) -> Tuple[str, ...]:
+        return tuple((Path(request.chdir) / path).as_posix() for path in paths)
+
     return Process(
-        argv=("__terraform/terraform",) + request.args,
+        argv=("__terraform/terraform", f"-chdir={shlex.quote(request.chdir)}") + request.args,
         input_digest=request.input_digest,
         immutable_input_digests=immutable_input_digests,
-        output_files=request.output_files,
-        output_directories=request.output_directories,
+        output_files=prepend_paths(request.output_files),
+        output_directories=prepend_paths(request.output_directories),
         description=request.description,
         level=LogLevel.DEBUG,
     )

--- a/src/python/pants/backend/terraform/tool.py
+++ b/src/python/pants/backend/terraform/tool.py
@@ -185,6 +185,7 @@ class TerraformProcess:
     description: str
     input_digest: Digest = EMPTY_DIGEST
     output_files: tuple[str, ...] = ()
+    output_directories: tuple[str, ...] = ()
 
 
 @rule
@@ -204,6 +205,7 @@ async def setup_terraform_process(
         input_digest=request.input_digest,
         immutable_input_digests=immutable_input_digests,
         output_files=request.output_files,
+        output_directories=request.output_directories,
         description=request.description,
         level=LogLevel.DEBUG,
     )

